### ISSUE TITLE
`configure_file` parser cannot handle certain constructions

### ISSUE THREAD

**Issue by amcn** [muon-build/muon#78]
As noted in #75, FVWM is currently experimenting with a meson build system and they have been trying to get muon to build it. 

muon's `configure_file` parser currently has difficulty handling one of their input files:

https://github.com/fvwmorg/fvwm3/blob/da6bff92cece15d5f24f993457835ae8d40b7a7e/bin/fvwm-menu-directory.in#L266-L272

Specifically the `/@/@@/` on line 270 confuses the parser.

**Comment by kzc** [comment]
@amcn Does this naïve `@@` workaround work for you?  

With the patch I can successfully run `muon setup -Dgolang=disabled build` locally on macos for the fvwm3 branch above, but I can't build it because I don't have the dependencies.

```patch
diff --git a/src/functions/kernel/configure_file.c b/src/functions/kernel/configure_file.c
index e1fa98c..b9a4fd0 100644
--- a/src/functions/kernel/configure_file.c
+++ b/src/functions/kernel/configure_file.c
@@ -280,7 +280,11 @@ write_mesondefine:
 				continue;
 			}
 
-			if (i <= id_start) {
+			if (i == id_start) {
+				// let's not consider @@ an error and skip over it
+				// https://github.com/muon-build/muon/issues/78
+				continue;
+			} else if (i < id_start) {
 				error_messagef(&src, id_location, log_error, "key of zero length not supported");
 				return false;
 			} else if (!obj_dict_index_strn(wk, dict, &src.src[id_start], i - id_start, &elem)) {

```
There's another instance of the `"key of zero length not supported"` error in the source file, but it was left alone because it wasn't exercised by this test case.

**Comment by kzc** [comment]
Patch above was updated to fix a bug.

**Comment by amcn** [comment]
Apologies for the late reply: I can confirm that with this patch, muon is able to complete the setup of FVWM3's meson branch.

I've run the muon test suite locally against the patch and there are no regressions.

With regards to the impact of the patch on the overall correctness of the configure_file parser, @annacrombie is the arbiter.

**Comment by kzc** [comment]
Here's an accompanying test patch for the new `configure_file` `@@` behavior:

```patch
diff --git a/common/127 generated assembly/meson.build b/common/127 generated assembly/meson.build
index 5c3ba2e..3fa181c 100644
--- a/common/127 generated assembly/meson.build	
+++ b/common/127 generated assembly/meson.build	
@@ -44,7 +44,7 @@ if cc.get_id() == 'clang-cl' and cc.version().version_compare('< 12.0.0') and cp
 endif
 
 if cc.symbols_have_underscore_prefix()
-    add_project_arguments('-DMESON_TEST__UNDERSCORE_SYMBOL', language: 'c')
+    add_project_arguments('-DMESON_TEST__UNDERSCORE_SYMBOL', language: ['c', 'assembly'])
 endif
 
 copy = find_program('copyfile.py')
diff --git a/common/14 configure file/config6.h.in b/common/14 configure file/config6.h.in
index 0a91542..958bdcd 100644
--- a/common/14 configure file/config6.h.in	
+++ b/common/14 configure file/config6.h.in	
@@ -38,3 +38,9 @@
 
 /* This should never happen in the real world, right? */
 #define MESSAGE12 "@var1@var2\\\\@var3@var4\\\\@"
+
+/* Leave @/@@ as is */
+#define MESSAGE13 "$name =~ s/@/@@/g;"
+
+/* Leave zero length @@ keys as is */
+#define MESSAGE14 "@:@@.@@@;@@@@ @@@@@"
diff --git a/common/14 configure file/prog6.c b/common/14 configure file/prog6.c
index b39f9da..d582353 100644
--- a/common/14 configure file/prog6.c	
+++ b/common/14 configure file/prog6.c	
@@ -13,5 +13,7 @@ int main(void) {
         || strcmp(MESSAGE9, "foovar2@")
         || strcmp(MESSAGE10, "foovar2\\@")
         || strcmp(MESSAGE11, "foobarbazqux")
-        || strcmp(MESSAGE12, "foovar2\\@var3@var4\\@");
+        || strcmp(MESSAGE12, "foovar2\\@var3@var4\\@")
+        || strcmp(MESSAGE13, "$name =~ s/@/@@/g;")
+        || strcmp(MESSAGE14, "@:@@.@@@;@@@@ @@@@@");
 }
diff --git a/common/178 bothlibraries/meson.build b/common/178 bothlibraries/meson.build
index e33507c..ab80c73 100644
--- a/common/178 bothlibraries/meson.build	
+++ b/common/178 bothlibraries/meson.build	
@@ -1,4 +1,9 @@
-project('both libraries linking test', 'c', 'cpp')
+project(
+    'both libraries linking test',
+    'c',
+    'cpp',
+    default_options: ['cpp_std=c++11'],
+)
 
 both_libs = both_libraries('mylib', 'libfile.c')
 dep = declare_dependency(link_with: both_libs)
```

A couple of unrelated previously failing tests were fixed to run on macos with Apple clang version 11.0.0 (clang-1100.0.33.17) when python3 was added to the PATH.

**Comment by kzc** [comment]
@annacrombie When I tried running https://github.com/muon-build/muon/commit/5abfa56f0b62c79c98d51eb0a8b2253c5280cb44 against the [test case above](https://github.com/muon-build/muon/issues/78#issuecomment-2357196258) it failed.

**Comment by kzc** [comment]
The fvwm3 branch in the top post fails to run `muon setup -Dgolang=disabled build` with 5abfa56f0b62c79c98d51eb0a8b2253c5280cb44.

**Comment by kzc** [comment]
Verified that fvwm3 `muon setup -Dgolang=disabled build` works with 226b7e1bca667827c6a8f3500b928daa2973e087.

Issue can be closed.

**Comment by amcn** [comment]
Closing after confirmation that the issue is resolved by 226b7e1.

@Kangie, thanks for bringing the underlying issues to our attention!
